tuhi/png.py

```diff
@@ -0,0 +1,87 @@
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 2 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+
+from gi.repository import GObject
+import cairo
+
+class JsonPng(GObject.Object):
+
+    def __init__(self, json, orientation, filename, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.json = json
+        self.timestamp = json['timestamp']
+        self.filename = filename
+        self.orientation = orientation.lower()
+        self._convert()
+
+    def _convert(self):
+        js = self.json
+        dimensions = js['dimensions']
+        if dimensions == [0, 0]:
+            width, height = 100, 100
+        else:
+            # Original dimensions are too big for SVG Standard
+            # so we normalize them
+            width, height = dimensions[0] // 100, dimensions[1] // 100
+
+        if self.orientation in ['portrait', 'reverse-portrait']:
+            size = (height, width)
+        else:
+            size = (width, height )
+
+        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, size[0], size[1])
+        ctx = cairo.Context(surface)
+        line_width = 3
+
+        # Paint a transparent background
+        ctx.set_source_rgba(0, 0, 0, 0)
+        ctx.paint()
+
+        ctx.set_antialias(cairo.Antialias.DEFAULT)
+        ctx.set_line_join(cairo.LINE_JOIN_ROUND)
+        ctx.set_source_rgb(0, 0, 0)
+
+        for stroke_num, s in enumerate(js['strokes']):
+
+            points_with_sk_width = []
+
+            for p in s['points']:
+
+                x, y = p['position']
+                # Normalize coordinates too
+                x, y = x / 100, y / 100
+
+                if self.orientation == 'reverse-portrait':
+                    x, y = y, width - x
+                elif self.orientation == 'portrait':
+                    x, y = height - y, x
+                elif self.orientation == 'reverse-landscape':
+                    x, y = width - x, height - y
+
+                # Pressure normalized range is [0, 0xffff]
+                delta = (p['pressure'] - 0x8000) / 0x8000
+                stroke_width = 3 + delta
+                points_with_sk_width.append((x, y, stroke_width))
+
+            for i, (x, y, stroke_width) in enumerate(points_with_sk_width):
+                x, y, stroke_width = points_with_sk_width[i]
+                ctx.set_line_width(stroke_width)
+
+                if i == 0:
+                    ctx.move_to(x, y)
+                else:
+                    ctx.line_to(x, y)
+
+            ctx.stroke()
+
+        surface.write_to_png(self.filename)
```

Comment:
I can't think of a much better way right now (other than callbacks and whatnot) but it very much seems that we *should* be sharing the code with the `JsonSvg`. This is 90% the same and will eventually become the source of minor discrepancies as we forget to fix.

Maybe a parent class for the converter that calls the interface to add points/strokes which is then handled in the filetype-specific subclass?